File: 3098-find-the-sum-of-subsequence-powers/3098-find-the-sum-of-subsequence-powers.java
class Solution {
    int mod = (int)1e9 + 7;
    
    private int findSumOfPowers(int index, int prevIndex, int currDiff, int[] nums, int k, Map<String, Integer> mp) {
        if (index == nums.length) return (k == 0) ? currDiff % mod : 0;

        String key = index + "$" + currDiff + "$" + k + "$" + prevIndex;
        if (mp.containsKey(key)) return mp.get(key);
     
        int not_take = findSumOfPowers(index + 1, prevIndex, currDiff, nums, k, mp) % mod;
          
        currDiff = (prevIndex == -1) ? currDiff : Math.min(currDiff, Math.abs(nums[index] - nums[prevIndex]));
            
        int take = findSumOfPowers(index + 1, index, currDiff, nums, k - 1, mp) % mod;
        
        int total = (take + not_take) % mod;
        mp.put(key, (int)(take + not_take) % mod);
        
        return total;
    }
    
    public int sumOfPowers(int[] nums, int k) {
        Arrays.sort(nums);
        
        Map<String, Integer> mp = new HashMap<>();
        
        return findSumOfPowers(0, -1, Integer.MAX_VALUE, nums, k, mp);
    }
}

File: 3098-find-the-sum-of-subsequence-powers/3098-find-the-sum-of-subsequence-powers.cpp
class Solution {
public:
    int mod = 1000000007;
    unordered_map<int, unordered_map<int, unordered_map<int, unordered_map<long, long>>>> dp;

    long solve(int ind, int last, int size, long long mn, vector<int> &nums, int k) {
        int n = nums.size();
        if (size == k) {
            return mn % mod;
        }
        if (ind >= n)
            return 0;

        if (dp[ind][last][size].count(mn))
            return dp[ind][last][size][mn];

        long pick = LONG_MAX, notPick = LONG_MAX;
        long long diff = LONG_MAX;
        if (last != -1) {
            diff = abs(nums[ind] * 1LL - nums[last]) % mod;
        }
        
        pick = solve(ind + 1, ind, size + 1, min(mn, diff), nums, k);

        notPick = solve(ind + 1, last, size, mn, nums, k);

        return dp[ind][last][size][mn] = (pick + notPick) % mod;

    }

    int sumOfPowers(vector<int> &nums, int k) {
        sort(nums.begin(), nums.end());

        return solve(0, -1, 0, LONG_MAX, nums, k);

    }
};
